src/language-markdown/print-whitespace.js

```diff
@@ -0,0 +1,317 @@
+  return isInSentenceWithCJSpaces(path);
+}
+
+/**
+ * @param {WordKind | undefined} kind
+ * @returns {boolean} `true` if `kind` is CJK (including punctuation marks)
+ */
+function isCJK(kind) {
+  return (
+    kind === KIND_CJ_LETTER ||
+    kind === KIND_K_LETTER ||
+    kind === KIND_CJK_PUNCTUATION
+  );
+}
+
+/**
+ * @param {WordKind | undefined} kind
+ * @returns {boolean} `true` if `kind` is letter (not CJK punctuation)
+ */
+function isLetter(kind) {
+  return (
+    kind === KIND_NON_CJK || kind === KIND_CJ_LETTER || kind === KIND_K_LETTER
+  );
+}
+
+/**
+ * @param {WordKind | undefined} kind
+ * @returns {boolean} `true` if `kind` is western or Korean letters (divides words by Space)
+ */
+function isWesternOrKoreanLetter(kind) {
+  return kind === KIND_NON_CJK || kind === KIND_K_LETTER;
+}
+
+/**
+ * Returns whether “whitespace” (`"" | " " | "\n"`; see `WhitespaceValue`) can converted to `"\n"`
+ *
+ * @param {AstPath} path
+ * @param {WhitespaceValue} value
+ * @param {*} options
+ * @param {AdjacentNodes | undefined} [adjacentNodes]
+ * @returns {boolean | "trueIfSpace"} `true` if “whitespace” can be converted to `"\n"`; `trueIfSpace` equals to true only if `canBeConvertedToSpace` returns `true`
+ */
+function isBreakable(path, value, options, adjacentNodes) {
+  if (
+    options.proseWrap !== "always" ||
+    getAncestorNode(path, SINGLE_LINE_NODE_TYPES)
+  ) {
+    return false;
+  }
+
+  if (
+    adjacentNodes === undefined ||
+    // Space are always breakable
+    value === " "
+  ) {
+    return true;
+  }
+
+  // Simulates latin words; see #6516 (https://github.com/prettier/prettier/issues/6516)
+  // [Latin][""][Hangul] & vice versa => Don't break
+  // [han & kana][""][hangul], either
+  if (
+    value === "" &&
+    ((adjacentNodes.previous?.kind === KIND_K_LETTER &&
+      isLetter(adjacentNodes.next?.kind)) ||
+      (adjacentNodes.next?.kind === KIND_K_LETTER &&
+        isLetter(adjacentNodes.previous?.kind)))
+  ) {
+    return false;
+  }
+
+  // https://en.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages
+  const violatesCJKLineBreakingRule =
+    (adjacentNodes.next?.value !== undefined &&
+      noBreakBefore.has(adjacentNodes.next?.value?.[0])) ||
+    (adjacentNodes.previous?.value !== undefined &&
+      noBreakAfter.has(adjacentNodes.previous?.value?.at(-1)));
+
+  // Targets not only "" but also "\n"
+  // Intentional violation of the line breaking rules (e.g. “ル\nール守れ\n！”) tends to be “corrected” ("\n" -> "") by formatted with a large value of `printWidth`.
+  // Eventually, by reformatted with a smaller value of `printWidth` or because of a paragraph revision, the rules are going to be applied to the place that used to violate them.
+  // e.g. “シ\nョ\nートカ\nット\n！\n？” (completely violates the rules on purpose) --[printWidth = 6]-->“ショー\nトカッ\nト！？” --[s/^/こんなところで/]--> “こんな\nところ\nで\nショー\nトカッ\nト！？” (completely complies to the rules)
```

Comment:
This comment is outdated, right?

Reply:
@tats-u Please go through all the comments in that file and check if they aren't outdated. After that we'll be good to merge.

Reply:
All comments are LGTM as of the latest commit.